### PR TITLE
Fix last login at

### LINKED ISSUE #40: Fix last login at
Last login at currently reflects the creation time of the most recent session. Most of time the value is not what developer would like to have.

- Use two field per user: `login_at` and `last_login_at`
    - `login_at`: the creation time of most recent session
    - `last_login_at`: the creation time of the second-most recent session
- When creating session:
    1. Set `last_login_at` = `login_at`
    2. Set `login_at` = session creation time = now

### PR DESCRIPTION
**Pull request by kiootic**
fix #40

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    volumes:
+      - redis_data:/redis_data
+    ports:
+      - "6379:6379"
+
+  authgear:
+    build: .
+    environment:
+      DEV_MODE: "true"
+      LOG_LEVEL: "debug"
```
Comment (louischan-oursky) [inline docker-compose.yaml:24]: docker-compose by default load `.env` so it is unnecessary.

File: pkg/auth/dependency/user/store.go [hunk 2]
```diff
@@ -113,16 +116,24 @@ func (s *Store) UpdateMetadata(userID string, metadata map[string]interface{}, u
...
-func (s *Store) UpdateLoginTime(userID string, lastLoginAt time.Time) error {
+func (s *Store) UpdateLoginTime(userID string, loginAt time.Time) (*time.Time, error) {
 	builder := s.SQLBuilder.Tenant().
 		Update(s.SQLBuilder.FullTableName("user")).
-		Set("last_login_at", lastLoginAt).
-		Where("id = ?", userID)
+		Set("last_login_at", squirrel.Expr("login_at")).
+		Set("login_at", loginAt).
+		Where("id = ?", userID).
+		Suffix("RETURNING last_login_at")
```
Comment (louischan-oursky) [inline pkg/auth/dependency/user/store.go:125]: I think we should add LoginAt and LastLoginAt to User. This function keeps returning error only. The caller calls Get to retrieve the updated user.

File: pkg/auth/dependency/user/commands_raw.go [hunk 3]
```diff
@@ -55,11 +55,12 @@ func (c *RawCommands) UpdateMetadata(user *model.User, metadata map[string]inter
...
-func (c *RawCommands) UpdateLoginTime(user *model.User, lastLoginAt time.Time) error {
-	if err := c.Store.UpdateLoginTime(user.ID, lastLoginAt); err != nil {
+func (c *RawCommands) UpdateLoginTime(user *model.User, loginAt time.Time) error {
+	lastLoginAt, err := c.Store.UpdateLoginTime(user.ID, loginAt)
+	if err != nil {
 		return err
 	}
 
-	user.LastLoginAt = &lastLoginAt
+	user.LastLoginAt = lastLoginAt
```
Comment (louischan-oursky) [inline pkg/auth/dependency/user/commands_raw.go:64]: Refetch the user here.

File: docker-compose.yaml (hunk 1 above)
Comment (kiootic) [reply docker-compose.yaml:24]: It loads .env to substitute variable references in docker-compose.yaml, but for container environment, either `environment` or `env_file` should be used.

File: docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    volumes:
+      - ./authgear.yaml:/authgear.yaml
+      - ./authgear.secrets.yaml:/authgear.secrets.yaml
```
Comment (louischan-oursky) [inline docker-compose.yaml:27]: The WORKDIR is `/app` so we should mount them as `/app/authgear.yaml` and `/app/authgear.secrets.yaml`.

File: docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    environment:
+      POSTGRES_USER: "postgres"
+      POSTGRES_PASSWORD: "postgres"
+    ports:
+      - "5432:5432"
+
+  redis:
+    image: redis:5.0
+    volumes:
+      - redis_data:/redis_data
```
Comment (louischan-oursky) [inline docker-compose.yaml:16]: Should be `- redis_data:/data`

Comment (kiootic) [reply docker-compose.yaml]: fixed.

**Pull request merged**